### ISSUE TITLE
Toy models in pyRiemann

### ISSUE THREAD

**Issue by plcrodrigues** [pyRiemann/pyRiemann#138]
Hi, everyone.

In this issue, I would like to invite you to discuss the idea of including simple **toy models** to `pyriemann`. 

I think this would be an important addition to the package because:

- It would help us **showcasing** classifier performance with a dataset for which we have total control. Although publicly available BCI datasets are cool, it is hard to really understand the nuts and bolts of an algorithm without a total control over a simplified model.
- It would be useful for **validating** and **illustrating** future contributions with new methods. For instance, I believe it is an important step towards including transfer learning methods (e.g. Riemannian Procrustes analysis) into `pyriemann`.
- It would serve as a **standard reference** to many papers in the literature. For instance, it would avoid seeing a new slightly different toy model being proposed every time one wants to validate its own brand new method in the SPD manifold.
- It would (?) be useful for the design of **unitary tests**.

Having said that, I would like to propose a first idea towards this goal. 

I've been playing for a while with the concept of Gaussian Riemannian distributions, as defined in here [[1]](https://arxiv.org/abs/1507.01760). The idea is to define a probability distribution function parametrized by a mean <img src="https://render.githubusercontent.com/render/math?math=\bar{Y}"> and a dispersion <img src="https://render.githubusercontent.com/render/math?math=\sigma"> such that

<img src="https://render.githubusercontent.com/render/math?math=\Large p(Y) = \frac{1}{\zeta(\sigma)}\exp\left(-\frac{\delta_R^2(Y, \bar{Y})}{2\sigma^2}\right)">

where <img src="https://render.githubusercontent.com/render/math?math=\zeta(\sigma)"> is a normalizing constant and <img src="https://render.githubusercontent.com/render/math?math=\delta_R"> is our usual AIRM distance between two SPD matrices.

Although the equation above looks simple, there is quite a lot of intricate maths in [[1]](https://arxiv.org/abs/1507.01760) in order to derive a sampling procedure from this distribution. It is based on the idea of spectrally decomposing a SPD matrix and sampling separately its strictly positive eigenvalues and orthonormal eigenvectors. The pdf for each of these factors are quite ugly but the authors of the article did a great job in deriving them. 

I've **coded the sampling scheme** presented in the paper and have obtained quite nice results. 

---
### Example 1

For instance, in the figure below I've used `pyriemann.embedding` to see how a dataset generated with a Riemannian Gaussian distribution looks like for different choices of parameters. 

<img src="https://user-images.githubusercontent.com/4588557/132105396-2110fd4c-4b62-4b11-85d1-65415f3e79ee.png" width="80%">

---
### Example 2

I have also considered a setup with two classes of <img src="https://render.githubusercontent.com/render/math?math=m \times m"> SPD matrices: one generated by a distribution centered at <img src="https://render.githubusercontent.com/render/math?math=\bar{Y}"> and dispersion <img src="https://render.githubusercontent.com/render/math?math=\sigma">, and the other centered at <img src="https://render.githubusercontent.com/render/math?math=\varepsilon \bar{Y}"> with the same dispersion (<img src="https://render.githubusercontent.com/render/math?math=\varepsilon > 0">). Note that the actual distance between the two class means is <img src="https://render.githubusercontent.com/render/math?math=\Delta = m\log^2(\varepsilon)">, which can be used as a proxy for the separability of the classes. To illustrate this, I've used the MDM classifier in `pyriemann.classification` to classify several simulated datasets in which the value of <img src="https://render.githubusercontent.com/render/math?math=\varepsilon"> are different. The figure below reports the 5-fold cross validated scores for different choices of <img src="https://render.githubusercontent.com/render/math?math=\sigma"> and with <img src="https://render.githubusercontent.com/render/math?math=m = 2">. 

<img src="https://user-images.githubusercontent.com/4588557/132105450-fadd7ec2-dcd2-4b07-ac17-26586860ec3d.png" width="80%">

---

Well, that's it for now. Would you guys have other suggestions of illustrations that could be useful? I can make a PR with the code and the examples if you think this is the way to go. Otherwise, I'm all ears to other suggestions.

Cheers,
Pedro

**Comment by agramfort** [comment]
I like that !

it reminds me the make_classification in sklearn.datasets submodule

https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_classification.html#sklearn.datasets.make_classification

>

**Comment by sylvchev** [comment]
Yes, this is a very good idea! I completely agree about the benefit for showcasing, examples and unit test. While reformatting the tests, this was really frustrating to be limited to generic covariance matrices without any control of the distribution. 

The examples are very impressive!

Also, it could be nice to have the bayesian classifier defined in the same papers. Do you think the code of the sampling could be reused for the classifier?
